src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -547,42 +525,93 @@ private static String toLowerCase(String s) {
         return new String(chars);
     }
 
-    /**
-     * Optionally decodes a String, if it's encoded. If String is not encoded,
-     * method will return the original input value.
-     *
-     * @param input the value String to decode
-     * @return a decoded String
-     */
-    private String percentDecode(final String input) {
-        if (input == null) {
-            return null;
-        }
-        final String decoded = uriDecode(input);
-        if (!decoded.equals(input)) {
-            return decoded;
+    private static String percentDecode(final String source) {
+        if (source == null || source.isEmpty()) {
+            return source;
         }
-        return input;
-    }
 
-    public static String uriDecode(String source) {
-        if (source == null) {
+        byte[] bytes = source.getBytes(StandardCharsets.UTF_8);
+        int percentCharCount = getPercentCharCount(bytes);
+
+        if (percentCharCount == 0) {
             return source;
         }
-        int length = source.length();
-        StringBuilder builder = new StringBuilder();
+
+        int length = bytes.length;
+        int capacity = (length + percentCharCount) - (percentCharCount * 3);
+
+        if (capacity <= 0) {
+            throw new ValidationException("Invalid encoding in '" + source + "'");
+        }
+
+        ByteBuffer buffer = ByteBuffer.allocate(capacity);
+
         for (int i = 0; i < length; i++) {
-            if (source.charAt(i) == '%') {
-                String str = source.substring(i + 1, i + 3);
-                char c = (char) Integer.parseInt(str, 16);
-                builder.append(c);
-                i += 2;
+            if (buffer.position() + 1 > capacity) {
+                throw new ValidationException("Invalid encoding in '" + source + "'");
             }
-            else {
-                builder.append(source.charAt(i));
+
+            int b;
+
+            if (bytes[i] == PERCENT_CHAR) {
+                int b1 = Character.digit(bytes[++i], 16);
+                int b2 = Character.digit(bytes[++i], 16);
+                b = (byte) ((b1 << 4) + b2);
```

Comment:
I still believe that checking if `b1` or `b2` is `-1` here does not hurt…

Reply:
Do we have a test case? I know it's possible for `Character::digit` to return `-1`, but _when_ will it in this case?

Reply:
`uriDecode("%GH")`? The method is deprecated, but still `public`.

It is not at all evident here, that `bytes` must only contain valid percent sequences. This is true, if this method is called from `parse` but not at all evident, so it merits at least a comment. Besides: this method already contains two `throw` statements.